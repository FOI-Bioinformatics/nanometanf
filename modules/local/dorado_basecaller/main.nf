process DORADO_BASECALLER {
    tag "$meta.id"
    label 'process_high'

    // Assume dorado is available in PATH
    // conda "${moduleDir}/environment.yml"

    input:
    tuple val(meta), path(pod5_files)
    val model

    output:
    tuple val(meta), path("*.fastq.gz"), emit: fastq
    tuple val(meta), path("*_summary.txt"), emit: summary
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def min_qscore = params.min_qscore ?: 9
    def trim_adapters = params.trim_adapters ? '--trim adapters' : ''
    def trim_barcodes = params.trim_barcodes ? '--trim' : ''
    def demultiplex = params.demultiplex && params.barcode_kit ? "--kit-name ${params.barcode_kit}" : ''

    // Handle both single POD5 file and directory of POD5 files
    def input_path = pod5_files.size() == 1 && pod5_files[0].isFile() ? pod5_files[0] : '.'
    
    """
    # Check if dorado is available
    if ! command -v dorado &> /dev/null; then
        echo "ERROR: Dorado not found in PATH"
        exit 1
    fi

    # GPU Detection and Device Selection
    echo "=== GPU Detection ==="
    GPU_AVAILABLE=false
    GPU_DEVICES=""
    
    # Check for NVIDIA GPUs
    if command -v nvidia-smi &> /dev/null; then
        echo "NVIDIA drivers detected, checking for available GPUs..."
        if nvidia-smi -L 2>/dev/null | grep -q "GPU"; then
            GPU_COUNT=\$(nvidia-smi -L | wc -l)
            echo "Found \$GPU_COUNT NVIDIA GPU(s)"
            GPU_AVAILABLE=true
            # Use all available GPUs
            GPU_DEVICES="cuda:all"
        else
            echo "NVIDIA drivers found but no GPUs detected"
        fi
    else
        echo "No NVIDIA drivers detected"
    fi
    
    # Check for Apple Silicon GPUs (Metal Performance Shaders)
    if [[ "\$(uname)" == "Darwin" ]] && [[ "\$(uname -m)" == "arm64" ]]; then
        if system_profiler SPDisplaysDataType 2>/dev/null | grep -q "Apple"; then
            echo "Apple Silicon GPU detected"
            GPU_AVAILABLE=true
            GPU_DEVICES="metal"
        fi
    fi
    
    # Set device argument based on availability
    if [ "\$GPU_AVAILABLE" = true ]; then
        DEVICE_ARG="--device \$GPU_DEVICES"
        echo "Using GPU acceleration: \$GPU_DEVICES"
        
        # Optimize batch size for GPU
        if [[ "\$GPU_DEVICES" == "cuda:all" ]]; then
            BATCH_SIZE_ARG="--batchsize 384"  # Optimized for NVIDIA GPUs
            CHUNK_SIZE_ARG="--chunksize 10000"
        elif [[ "\$GPU_DEVICES" == "metal" ]]; then
            BATCH_SIZE_ARG="--batchsize 192"  # Conservative for Apple Silicon
            CHUNK_SIZE_ARG="--chunksize 4000"
        fi
    else
        DEVICE_ARG="--device cpu"
        BATCH_SIZE_ARG="--batchsize 128"   # Conservative for CPU
        CHUNK_SIZE_ARG="--chunksize 4000"
        echo "Using CPU processing (no suitable GPU found)"
    fi
    
    echo "Device configuration: \$DEVICE_ARG \$BATCH_SIZE_ARG \$CHUNK_SIZE_ARG"

    # Download model if not available locally  
    echo "=== Model Download ==="
    dorado download --model ${model} || echo "Model may already be available"

    # Run basecalling with optimal settings
    echo "=== Basecalling ==="
    dorado basecaller \\
        ${model} \\
        ${input_path} \\
        \$DEVICE_ARG \\
        \$BATCH_SIZE_ARG \\
        \$CHUNK_SIZE_ARG \\
        --emit-fastq \\
        --min-qscore ${min_qscore} \\
        --verbose \\
        ${trim_adapters} \\
        ${trim_barcodes} \\
        ${demultiplex} \\
        ${args} \\
        > ${prefix}.fastq

    # Compress output
    echo "=== Compression ==="
    gzip ${prefix}.fastq

    # Create detailed summary file
    echo "=== Summary Generation ==="
    cat > ${prefix}_summary.txt << EOF
Sample: ${prefix}
Model: ${model}
Min Q-score: ${min_qscore}
Device: \$GPU_DEVICES
GPU Available: \$GPU_AVAILABLE
Trim Adapters: ${params.trim_adapters ?: false}
Trim Barcodes: ${params.trim_barcodes ?: false}
Demultiplex: ${params.demultiplex ?: false}
Barcode Kit: ${params.barcode_kit ?: 'none'}
Input files: ${pod5_files.join(', ')}
Input file count: ${pod5_files.size()}
Basecalling started: \$(date)
Basecalling completed: \$(date)
Host: \$(hostname)
CPUs: ${task.cpus}
Memory: ${task.memory}
EOF

    # Add GPU information to summary if available
    if [ "\$GPU_AVAILABLE" = true ]; then
        echo "GPU Information:" >> ${prefix}_summary.txt
        if command -v nvidia-smi &> /dev/null; then
            nvidia-smi --query-gpu=name,memory.total,utilization.gpu --format=csv,noheader,nounits >> ${prefix}_summary.txt 2>/dev/null || echo "GPU info unavailable" >> ${prefix}_summary.txt
        elif [[ "\$(uname)" == "Darwin" ]]; then
            system_profiler SPDisplaysDataType | grep "Chipset Model" >> ${prefix}_summary.txt 2>/dev/null || echo "Apple GPU info" >> ${prefix}_summary.txt
        fi
    fi

    # Version information
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dorado: \$(dorado --version 2>&1 | head -n 1 | sed 's/.*dorado //g')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def min_qscore = params.min_qscore ?: 9
    """
    # Create stub FASTQ with realistic structure
    cat > ${prefix}.fastq << 'EOF'
@stub_read_001 runid=stub basecall_model_version=${model}
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
+
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
@stub_read_002 runid=stub basecall_model_version=${model}
TGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCA
+
JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ
EOF
    gzip ${prefix}.fastq

    # Create comprehensive summary matching real output structure
    cat > ${prefix}_summary.txt << EOF
Sample: ${prefix}
Model: ${model}
Min Q-score: ${min_qscore}
Device: cpu
GPU Available: false
Trim Adapters: ${params.trim_adapters ?: false}
Trim Barcodes: ${params.trim_barcodes ?: false}
Demultiplex: ${params.demultiplex ?: false}
Barcode Kit: ${params.barcode_kit ?: 'none'}
Input files: stub_input.pod5
Input file count: 1
Basecalling started: \$(date "+%Y-%m-%d %H:%M:%S")
Basecalling completed: \$(date "+%Y-%m-%d %H:%M:%S")
Host: \$(hostname)
CPUs: ${task.cpus}
Memory: ${task.memory}
EOF

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dorado: 1.1.1
    END_VERSIONS
    """
}