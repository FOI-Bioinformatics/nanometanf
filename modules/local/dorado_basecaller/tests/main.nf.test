nextflow_process {

    name "Test DORADO_BASECALLER"
    script "../main.nf"
    process "DORADO_BASECALLER"

    tag "modules"
    tag "modules_local"
    tag "dorado_basecaller"

    test("Should run dorado basecalling on POD5 files") {

        tag "basecalling"

        setup {
            """
            // Create test POD5 file structure (simulate with dummy files)
            mkdir -p $outputDir/test_pod5_input
            echo 'dummy pod5 content' > $outputDir/test_pod5_input/test.pod5
            echo 'dummy pod5 content 2' > $outputDir/test_pod5_input/test2.pod5
            """
        }

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'test_sample',
                        single_end: true,
                        basecalling_model: 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                    ],
                    file('$outputDir/test_pod5_input')
                ]
                input[1] = 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should run dorado basecalling with demultiplexing") {

        setup {
            """
            mkdir -p $outputDir/test_pod5_demux
            echo 'dummy pod5 content' > $outputDir/test_pod5_demux/test.pod5
            """
        }

        when {
            params {
                barcode_kit = 'SQK-RBK004'
                demultiplex = true
            }

            process {
                """
                input[0] = [
                    [
                        id: 'test_demux_sample',
                        single_end: true,
                        basecalling_model: 'dna_r10.4.1_e8.2_400bps_sup@v4.2.0'
                    ],
                    file('$outputDir/test_pod5_demux')
                ]
                input[1] = 'dna_r10.4.1_e8.2_400bps_sup@v4.2.0'
                """
            }
        }

        then {
            assert process.success
            assert process.out.reads.size() == 1
            
            // Check output structure with demultiplexing
            def (meta, reads) = process.out.reads[0]
            assert meta.id == 'test_demux_sample'
            
            // Should have versions
            assert process.out.versions.size() == 1
        }
    }

    test("Should handle different basecalling models") {

        setup {
            """
            mkdir -p $outputDir/test_pod5_models
            echo 'dummy pod5 content' > $outputDir/test_pod5_models/test.pod5
            """
        }

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'test_model_sample',
                        single_end: true,
                        basecalling_model: 'dna_r9.4.1_e8_fast@v3.4'
                    ],
                    file('$outputDir/test_pod5_models')
                ]
                input[1] = 'dna_r9.4.1_e8_fast@v3.4'
                """
            }
        }

        then {
            assert process.success
            assert process.out.reads.size() == 1
            
            def (meta, reads) = process.out.reads[0]
            assert meta.basecalling_model == 'dna_r9.4.1_e8_fast@v3.4'
            
            assert process.out.versions.size() == 1
        }
    }

    test("Should handle empty POD5 directory gracefully") {

        setup {
            """
            mkdir -p $outputDir/test_empty_pod5
            # Create empty directory
            """
        }

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'test_empty',
                        single_end: true,
                        basecalling_model: 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                    ],
                    file('$outputDir/test_empty_pod5')
                ]
                input[1] = 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                """
            }
        }

        then {
            // Should handle empty directory gracefully
            assert process.success || process.failed
            
            if (process.success) {
                // If successful, should still produce minimal output
                assert process.out.reads.size() >= 0
                assert process.out.versions.size() == 1
            }
        }
    }

    test("Should run in stub mode") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'test_stub',
                        single_end: true,
                        basecalling_model: 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                    ],
                    []
                ]
                input[1] = 'dna_r10.4.1_e8.2_400bps_hac@v4.2.0'
                """
            }
        }

        then {
            assert process.success
            assert process.out.reads.size() == 1
            assert process.out.versions.size() == 1
            
            // In stub mode, should produce placeholder files
            def (meta, reads) = process.out.reads[0]
            assert meta.id == 'test_stub'
        }
    }
}