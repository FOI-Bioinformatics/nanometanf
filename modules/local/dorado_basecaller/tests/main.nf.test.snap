{
    "Should run dorado basecalling on POD5 files": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "fastq": [
                    
                ],
                "summary": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T11:56:53.765529"
    }
}